serverless/azure/src/get.ts

```diff
@@ -0,0 +1,138 @@
+import { HttpResponseInit } from "@azure/functions";
+import { Headers } from "undici";
+import {
+  Compression,
+  PMTiles,
+  SharedPromiseCache,
+  Source,
+  TileType,
+} from "pmtiles";
+import { tileJSON } from "../../shared";
+
+class KeyNotFoundError extends Error {
+  constructor(message: string) {
+    super(message);
+  }
+}
+
+async function nativeDecompress(
+  buf: ArrayBuffer,
+  compression: Compression
+): Promise<ArrayBuffer> {
+  if (compression === Compression.None || compression === Compression.Unknown) {
+    return buf;
+  } else if (compression === Compression.Gzip) {
+    let stream = new Response(buf).body!;
+    let result = stream.pipeThrough(new DecompressionStream("gzip"));
+    return new Response(result).arrayBuffer();
+  } else {
+    throw Error("Compression method not supported");
+  }
+}
+
+//const CACHE = new ResolvedValueCache(25, undefined, nativeDecompress);
+const CACHE = new SharedPromiseCache(25, undefined, nativeDecompress);
+
+export async function getZxy(
+  source: Source,
+  tile: [number, number, number] | undefined,
+  ext: string,
+  allowed_origin: string,
+  hostname: string,
+  ifNonMatchEtag: string | null
+): Promise<HttpResponseInit> {
+  const cacheableResponse = (
+    body: ArrayBuffer | string | undefined,
+    cacheable_headers: Headers,
+    status: number
+  ) => {
+    let resp_headers = new Headers(cacheable_headers);
+    if (allowed_origin)
+      resp_headers.set("Access-Control-Allow-Origin", allowed_origin);
+    resp_headers.set("Vary", "Origin");
+    resp_headers.set(
+      "Cache-Control",
+      "max-age=" + (process.env.CACHE_MAX_AGE ?? 86400)
+    );
+    return { body, headers: resp_headers, status: status };
+  };
+
+  const cacheableHeaders = new Headers();
+
+  const p = new PMTiles(source, CACHE, nativeDecompress);
+  try {
+    const p_header = await p.getHeader();
+
+    if (
+      (ifNonMatchEtag && ifNonMatchEtag === p_header.etag) ||
```

Comment:
@tengl we are working through the ETag logic in https://github.com/protomaps/go-pmtiles/pull/137 and I think the behavior here may not be intended. While it's OK for all the individual tiles to have the same ETag as the whole archive as implemented here, we can't short-circuit the 304 response, because the underlying archive may have changed on disk; in other words every request to the Azure Function must make at least one request to the storage blob to properly reflect the present state of the ETag. 

Given that this optimization isn't possible any longer I suggest that we converge the PMTiles decoding implementations on generating content-based etags for each unique tile, like the Lambda function does now: https://github.com/protomaps/PMTiles/blob/main/serverless/aws/src/index.ts#L238

This would also provide the benefit that updating an archive does not invalidate all etags, any tiles that remain the same across archives can be cached by intermediate CDNs.

Would making this change to unique ETags still work for your deployment?

Reply:
Well, the Azure Function would always have to make a request to the storage in order to detect changes, right? Isn't that the point of ETags, the only thing they save is the data transfer. I don't understand how you can short-circuit the 304 response and still detect changes in the archive?

So the process is something like this? 

* Client request tile without ETag.
* Request tile data from archive.
* Store the ETag for the archive.
* Calculate and store ETag for the tile.
* Return tile ETag to client.
* New client request for same tile with ETag.
* Request tile data from archive with ETag for archive.
* No change, no data, return stored ETag for tile.
* UPDATE OF ARCHIVE.
* Client request same tile with ETag.
* Request tile data from archive, with ETag for archive.
* ETag mismatch, new data in response. 
* Overwrite ETag for archive.
* Calculate new ETag for tile. 
* Check if client ETag is same as new calculated ETag.
* If yes, return unchanged, if no, return data and new ETag.

That saves the data transfer between the Azure Function and the client, but not from the storage blob to the Azure Function.

I don't think we need to care for my deployment, I use my own repo so I can change the code however I want and it makes it easier to deploy using my current tooling.

Reply:
My understanding of the implementation on line 67 is that it short circuits based on the archive etag and the If-None-Match request header matching. but the call to `p.getHeader()` could be returning a cached, not fresh value. So on step 10-12 above, the azure function will short-circuit and not detect the mismatch at line 12. Am I interpreting that right or missing something? (it's been awhile since I've looked at this part of the code)

Reply:
No, you're right. We don't want that. 

But how would unique ETags work? Like the process I described above?

Reply:
for an Azure Function written in JavaScript, it would call one of the Node standard library hashing functions such as `md5` on the tile contents, then set those as a strong ETag header. So this would add some computational overhead to every request. Same as how the [lambda works here](https://github.com/protomaps/PMTiles/blob/main/serverless/aws/src/index.ts#L238)

Reply:
Yes, but the overhead would also be a data transfer from the storage to the function. Every request needs to calculate the ETag with the tile data, right? At least if we don't check the ETag of the archive. 

An optimized version just for HTTP (Fetch) sources could do a HTTP HEAD request at regular intervals to check if the ETag for the archive has changed. If it has, invalidate cache and load new data, otherwise short-circuit and return 304. The tiles could still have individual ETags, and they could still be the same after the archive has changed. That would save a data transfer from the function to the client if the ETag is still the same. But I don't know how that would be implemented and still fit other sources.

Reply:
I think these are two different perspectives on the purpose of the Azure Function:

1. It is a "view" into the storage, not a cache. So every request into the function generates at least one data transfer to the underlying storage. There is caching going on, but it is the minimal necessary parts (headers, directories) to accomplish efficient behavior. It is paired with a standard HTTP cache in front of it like Azure CDN or CloudFront. 
2. It itself acts like a cache and is allowed to return stale data. In this case you can optimize away some requests by using the header ETag, at the cost of possibly returning a version that is older than the underlying storage.

Reply:
1. Yes, I agree that it is a view and that it should cache as little as possible. Leaving all caching to CDN is fine. 
2. Sure, depending on how long the ETag for the archive is valid but I suspect the CDN will return stale data more often (depending on Cache-Control values, I guess). Keeping the function simple it probably better than optimizing for some cases where the CDN doesn't hit the cache.  

Anyway, the function is pretty fast, normally about 200-300 ms round-trip time but it can take up to 700 ms. From my very rudimentary tests, I think that at least half of that time is between the CDN and the function, and the rest is between the function and storage. 

My subjective impression is that the map feels pretty responsive even when navigating in areas that the CDN hasn't cached. So the conclusion is:

1. Keep the function as simple as possible, translate the incoming request and fetch the data from the storage.
2. Recalculate the ETag from the tile content. 
3. Avoid caching. 
4. I suggest the configuration of the function only needs an URL to a file (with SAS token), which means that most of the code can be used in other hosting environments that has HTTP hosted archives.